### PR TITLE
[EmitC][NFC] Introduce common baseclass for conversions and simplify pattern insertion

### PR DESCRIPTION
**Pull request by simon-camp**
This makes future refactoring like adding constructor arguments to conversions easier.

Additionally pattern insertion is now sorted alphabetically per macro group, as the ordering by extension wasn't fully followed anyhow.

### CODE REVIEW

**Comment by simon-camp** [comment]
In preparation for #14702

**Review by benvanik — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 1]
```diff
@@ -1243,22 +1243,33 @@ void updateResultUses(Operation *op, ConversionPatternRewriter &rewriter,
...
+template <typename OpTy>
+class EmitCConversionPattern : public OpConversionPattern<OpTy> {
+public:
+  using Adaptor = typename OpTy::Adaptor;
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+
+  EmitCConversionPattern(const TypeConverter &typeConverter,
+                         MLIRContext *context)
+      : OpConversionPattern<OpTy>(typeConverter, context) {}
+};
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp:1255]: Something in here is ambiguous, breaking builds on Windows / MSVC

https://github.com/openxla/iree/actions/runs/6614754015/job/17965409798#step:8:7604

```
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\memory(3393): error C2668: 'mlir::iree_compiler::`anonymous-namespace'::BranchOpConversion::BranchOpConversion': ambiguous call to overloaded function
C:\a\iree\iree\compiler\src\iree\compiler\Dialect\VM\Conversion\VMToEmitC\ConvertVMToEmitC.cpp(2878): note: could be 'mlir::iree_compiler::`anonymous-namespace'::BranchOpConversion::BranchOpConversion(const mlir::TypeConverter &,mlir::MLIRContext *)', which inherits 'mlir::iree_compiler::`anonymous-namespace'::EmitCConversionPattern<mlir::iree_compiler::IREE::VM::BranchOp>::EmitCConversionPattern(const mlir::TypeConverter &,mlir::MLIRContext *)' via base class 'mlir::iree_compiler::`anonymous-namespace'::EmitCConversionPattern<mlir::iree_compiler::IREE::VM::BranchOp>'
C:\a\iree\iree\compiler\src\iree\compiler\Dialect\VM\Conversion\VMToEmitC\ConvertVMToEmitC.cpp(2878): note: or       'mlir::iree_compiler::`anonymous-namespace'::BranchOpConversion::BranchOpConversion(const mlir::TypeConverter &,mlir::MLIRContext *,mlir::PatternBenefit)', which inherits 'mlir::OpConversionPattern<OpTy>::OpConversionPattern(const mlir::TypeConverter &,mlir::MLIRContext *,mlir::PatternBenefit)' via base class 'mlir::iree_compiler::`anonymous-namespace'::EmitCConversionPattern<mlir::iree_compiler::IREE::VM::BranchOp>'
        with
        [
            OpTy=mlir::iree_compiler::IREE::VM::BranchOp
        ]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\memory(3393): note: while trying to match the argument list '(mlir::iree_compiler::IREE::VM::EmitCTypeConverter, mlir::MLIRContext *)'
C:\a\iree\iree\third_party\llvm-project\mlir\include\mlir/IR/PatternMatch.h(275): note: see reference to function template instantiation 'std::unique_ptr<T,std::default_delete<T>> std::make_unique<T,mlir::iree_compiler::IREE::VM::EmitCTypeConverter&,mlir::MLIRContext*&,0>(mlir::iree_compiler::IREE::VM::EmitCTypeConverter &,mlir::MLIRContext *&)' being compiled
        with
        [
            T=mlir::iree_compiler::`anonymous-namespace'::BranchOpConversion
        ]
C:\a\iree\iree\third_party\llvm-project\mlir\include\mlir/IR/PatternMatch.h(1850): note: see reference to function template instantiation 'std::unique_ptr<T,std::default_delete<T>> mlir::RewritePattern::create<T,mlir::iree_compiler::IREE::VM::EmitCTypeConverter&,mlir::MLIRContext*&>(mlir::iree_compiler::IREE::VM::EmitCTypeConverter &,mlir::MLIRContext *&)' being compiled
        with
        [
            T=mlir::iree_compiler::`anonymous-namespace'::BranchOpConversion
        ]
```

Revert then fix-forward? Or is there a simple fix?

Comment (simon-camp) [reply compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp:1255]: Feel free to revert, it's getting late here. I will look into this tomorrow.

Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp:1255]: Ok, thanks. I may take a look myself once I can switch workspaces.

Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp:1255]: Sent https://github.com/openxla/iree/pull/15267

**Pull request merged**